packages/teams-js/src/public/app.ts

```diff
@@ -806,6 +816,100 @@ export namespace app {
       resolve(sendAndHandleStatusAndReason('executeDeepLink', deepLink));
     });
   }
+
+  /**
+   * Namespace to opt into full-caching or delayed-termination of an app.
+   *
+   * @beta
+   */
+  export namespace lifecycle {
+    /**
+     * Register on resume handler function type
+     *
+     * @param context - Data structure to be used to pass the context to the app.
+     */
+    type registerOnResumeHandlerFunctionType = (context: ResumeContext) => void;
+    /**
+     * A function that is called when the app is ready to be suspended or terminated.
+     */
+    type ReadyToSuspendOrTerminateFunction = () => void;
+
+    /**
+     * Register before suspendOrTerminate handler function type
+     *
+     * @param readyToSuspendOrTerminate - A function that the handler is called with. Apps should call this function when they are ready.
+     * @returns void
+     */
+    type registerBeforeSuspendOrTerminateHandlerFunctionType = (
+      readyToSuspendOrTerminate: ReadyToSuspendOrTerminateFunction,
+    ) => void;
+
+    /**
+     * Registers a handler to be called before the page is suspended or terminatd.
+     *
+     * @param handler - The handler to invoke before the page is suspended or terminated. The page should
+     * invoke the readyToSuspendOrTerminate function provided to it once it's ready.
+     *
+     */
+    export function registerBeforeSuspendOrTerminateHandler(
+      handler: registerBeforeSuspendOrTerminateHandlerFunctionType,
+    ): void {
+      handler && ensureInitialized(runtime);
```

Comment:
>handler && ensureInitialized(runtime); [](http://example.com/codeflow?start=6&length=38)

I think I might know the answer to this but why do we have to do this here? Doesn't `ensureInitialized` get called from within `isSupported` ?

Reply:
I am not sure the reason, but we have been calling the ensureInitialized and isSuuported for all (afaik) APIs. @TrevorJoelHarris  can you help us please?

Reply:
I talked to Trevor, and the reason is,  

Hosts don't alter what they say are supported in the runtime object based on the current framecontext. Because of this, we have some capabilities where different combinations of functions work depending on what the frameContext is. 

For example, dialog.url.open, hosts always say they support dialog.url, no matter what the framecontext is, but if the app is in the tasks context, we don't want that function to work because the app shouldn't be able to make a dialog that opens another dialog.

Reply:
I can see how that would apply if the parameters being passed to `ensureInitialized` were different (e.g., `dialog.bot.open` passes three frame contexts and `dialog.bot.isSupported` passes none). But at least here, the same frame contexts are being passed in `registerBeforeSuspendOrTerminateHandler` and in `isSupported`?

This isn't a big deal if you want to leave it this way.

Reply:
deleted the ensure initialized call :)